core/src/main/java/io/grpc/transport/ServerTransportListener.java

```diff
@@ -48,4 +49,9 @@
    */
   ServerStreamListener streamCreated(ServerStream stream, String method,
       Metadata.Headers headers);
+
+  /**
+   * The transport completed shutting down. All resources have been released.
+   */
+  void transportTerminated();
```

Comment:
So I have two main problems with a single state as an API:

1. It doesn't answer the question users will be asking. This means instead of doing ```if (isShutdown())``` like you do now, you have to do ```State state = state(); if (state == SHUTTING_DOWN || state == TERMINATED)```. Note that if we had a FAILED case, like Guava's Service, you would have to include it as well. This is annoying.

2. You can't add new states in the future, because of point 1. That means you either error on the side of making too many states (making point 1 worse) or you error on the side of too few and now you have the same problem we have now. This is quite worrisome.

So the API is annoying to use and non-extensible. That seems to be the opposite of a "good API".

The benefit of a state() API is that it has a single method and that method gets you the state atomically. Having one method instead of two isn't very much of a savings. And we don't actually need the _states_. Other APIs do need to expose concrete states, but ours doesn't, so this possible benefit doesn't apply to us. (And generally such APIs don't suffer from point 1 above, so point 2 is also not an issue.)